execution/stream_join.go

```diff
@@ -0,0 +1,448 @@
+	if updatedRecordCount.AsInt() == 0 {
+		err = recordCountState.Clear()
+		if err != nil {
+			return 0, errors.Wrap(err, "couldn't clear count state of record")
+		}
+	} else {
+		err = recordCountState.Set(&updatedRecordCount)
+		if err != nil {
+			return 0, errors.Wrap(err, "couldn't update the count of record")
+		}
+	}
+
+	return updatedRecordCount.AsInt(), nil
+}
+
+func recordToValue(rec *Record) octosql.Value {
+	fields := make([]octosql.Value, len(rec.FieldNames))
+	data := make([]octosql.Value, len(rec.Data))
+	eventTimeField := octosql.MakeString(rec.EventTimeField().String())
+
+	for i := range rec.FieldNames {
+		fields[i] = octosql.MakeString(rec.FieldNames[i])
+		data[i] = *rec.Data[i]
+	}
+
+	valuesTogether := make([]octosql.Value, 4)
+
+	valuesTogether[0] = octosql.MakeTuple(fields)
+	valuesTogether[1] = octosql.MakeTuple(data)
+	valuesTogether[2] = octosql.MakeBool(rec.IsUndo())
+	valuesTogether[3] = eventTimeField
+
+	return octosql.MakeTuple(valuesTogether)
+}
+
+func valueToRecord(val octosql.Value) *Record {
+	slice := val.AsSlice()
+
+	fieldNameValues := slice[0].AsSlice()
+	values := slice[1].AsSlice()
+	isUndo := slice[2].AsBool()
+	eventTimeField := octosql.NewVariableName(slice[3].AsString())
+
+	fieldNames := make([]octosql.VariableName, len(fieldNameValues))
+
+	for i, fieldName := range fieldNameValues {
+		fieldNames[i] = octosql.NewVariableName(fieldName.AsString())
+	}
+
+	if isUndo {
+		return NewRecordFromSlice(fieldNames, values, WithEventTimeField(eventTimeField), WithUndo())
+	}
+
+	return NewRecordFromSlice(fieldNames, values, WithEventTimeField(eventTimeField), WithNoUndo())
+}
+
+// We create a octosql.Value from two records (as values), a flag whether the match is a retraction,
+// and an int indicating which of the records came later
+func createMatch(first, second octosql.Value, isUndo bool) octosql.Value {
+	return octosql.MakeTuple([]octosql.Value{first, second, octosql.MakeBool(isUndo)})
+}
+
+func unwrapMatch(match octosql.Value) (octosql.Value, octosql.Value, bool) {
+	asSlice := match.AsSlice()
+	return asSlice[0], asSlice[1], asSlice[2].AsBool()
+}
+
+func mergeRecords(left, right *Record, ID *RecordID, isUndo bool, eventTimeField octosql.VariableName) *Record {
+	mergedFieldNames := octosql.StringsToVariableNames(append(left.FieldNames, right.FieldNames...))
+	mergedDataPointers := append(left.Data, right.Data...)
+
+	mergedData := make([]octosql.Value, len(mergedDataPointers))
+
+	for i := range mergedDataPointers {
+		mergedData[i] = *mergedDataPointers[i]
+	}
+
+	// If eventTimeField is not empty, then we are joining by event_time and thus both records have the same value
+	if eventTimeField != "" {
+		eventTime := left.EventTime()
+		mergedFieldNames = append(mergedFieldNames, eventTimeField)
+		mergedData = append(mergedData, eventTime)
+	}
+
+	if isUndo {
+		return NewRecordFromSlice(mergedFieldNames, mergedData, WithID(ID), WithUndo(), WithEventTimeField(eventTimeField))
+	} else {
```

Comment:
You can keep one pattern for returns - just like that or like in `valueToRecord`